src/java/org/apache/cassandra/db/tries/MemtableTrie.java

```diff
@@ -0,0 +1,1039 @@
+        }
+
+        /**
+         * After a node's children are processed, this is called to ascend from it. This means applying the collected
+         * content to the compiled updatedPostContentNode and creating a mapping in the parent to it (or updating if
+         * one already exists).
+         * Returns true if still have work to do, false if the operation is completed.
+         */
+        private boolean attachAndMoveToParentState() throws SpaceExhaustedException
+        {
+            int updatedPreContentNode = applyContent();
+            int existingPreContentNode = existingPreContentNode();
+            --currentDepth;
+            if (currentDepth == -1)
+            {
+                assert root == existingPreContentNode;
+                if (updatedPreContentNode != existingPreContentNode)
+                {
+                    // Only write to root if they are different (value doesn't change, but
+                    // we don't want to invalidate the value in other cores' caches unnecessarily).
+                    root = updatedPreContentNode;
+                }
+                return false;
+            }
+            if (updatedPreContentNode != existingPreContentNode)
+                attachChild(transition(), updatedPreContentNode);
+            return true;
+        }
+    }
+
+    /**
+     * Somewhat similar to {@link MergeResolver}, this encapsulates logic to be applied whenever new content is being
+     * upserted into a {@link MemtableTrie}. Unlike {@link MergeResolver}, {@link UpsertTransformer} will be applied no
+     * matter if there's pre-existing content for that trie key/path or not.
+     *
+     * @param <T> The content type for this {@link MemtableTrie}.
+     * @param <U> The type of the new content being applied to this {@link MemtableTrie}.
+     */
+    public interface UpsertTransformer<T, U>
+    {
+        /**
+         * Called when there's content in the updating trie.
+         *
+         * @param existing Existing content for this key, or null if there isn't any.
+         * @param update   The update, always non-null.
+         * @return The combined value to use.
+         */
+        T apply(T existing, U update);
+    }
+
+    /**
+     * Modify this trie to apply the mutation given in the form of a trie. Any content in the mutation will be resolved
+     * with the given function before being placed in this trie (even if there's no pre-existing content in this trie).
+     * @param mutation the mutation to be applied, given in the form of a trie. Note that its content can be of type
+     * different than the element type for this memtable trie.
+     * @param transformer a function applied to the potentially pre-existing value for the given key, and the new
+     * value. Applied even if there's no pre-existing value in the memtable trie.
+     */
+    public <U> void apply(Trie<U> mutation, final UpsertTransformer<T, U> transformer) throws SpaceExhaustedException
```

Comment:
Given the commentary about writes needing to be atomic in the class JavaDoc, would it make sense to have the first argument be explicitly typed as a `SingletonTrie` for now?